### ISSUE TITLE
Separate build tools for users from build tools used by Homebrew

### ISSUE THREAD

**Issue by carlmjohnson** [Homebrew/brew#14749]
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

I propose that for formula that are build tools such as Go, Python, Ruby, Node, etc. instead of having one formula which then is used to build, e.g., Go and also build all of the other formulas that rely on Go, there should be two formulas, one which is intended for end users to use to build their own software, and one which is intended for Homebrew to build a coherent set of packages from.

### What is the motivation for the feature?

Homebrew is a distribution manager. It tries to create a coherent set of packages that can be built together and work together somewhat coherently. That is a lot of work, and I for one thank all the volunteers who work on it. However, users frequently don't appreciate this aspect of Homebrew. They think of it as an installer, and they get confused why they can't just install the latest release of Go or Python or Node or whatever as soon as it comes out. The reason is that while Homebrew can just install Go 1.X more or less as soon as it exists, _all of the existing formulas that rely on Go need to be updated to work with the new release first_, and this can take a considerable amount of time (for example as of this writing, it has taken 21 days with no new release). This is a constant source of user confusion and endless whining in the issues from people who want their build tool to be released sooner. While I sympathize with the maintainers who have to tune out this whining, I feel that the root of this problem is a mismatch of expectations, and one cannot expect the users to change. Users want what they want. The best way to move forward is to just give it to them in the form of separate formulas for build tools that they are intended to use from the ones that Homebrew itself uses.

This also frequently comes up with Python, where users are confused that upgrading one formula that happens to use Python also ends up upgrade Python itself and breaking any virtualenvs that relied on a fixed version of Python being there. I'm sure there are many more examples of this that can be multiplied.

### How will the feature be relevant to at least 90% of Homebrew users?

It will bring the user's mental model of how Homebrew works closer to the reality of how it actually works, leading to less whining in the issue and better sanity for everyone involved.

### What alternatives to the feature have been considered?

I'm sorry if I'm filing this in the wrong place, or if it has been discussed to death elsewhere, but I've been waiting for the Go 1.20 formula to land and watching the comments, and it shows a persistent mismatch between the point of view of Brew developers and Brew users that I'd like to try to explain and suggest a solution for.

**Comment by MikeMcQuaid** [comment]
Thanks for the issue and the kind words @carlmjohnson.

What we could do more of, and probably should, is land a `go@1.20` formula the day it gets released and then migrate over from Go 1.19 to Go 1.20 at a later point. Then users could install the version of Go they want immediately while not being restricted on Homebrew's versions.

Alternatively, we could make more formulae which end up blocked on updates and have people complaining (like Go, Rust, etc.) always use `depends_on "go@1.X"` which makes incremental migration and adding a new formula easier.

We've done this for Python and it's made some things easier and some things harder.

Note that your proposal of having `go-homebrew` and `go` would just get complaints of "why do I need to have multiple versions of Go installed?" for anyone using tools that also depend on Go.

CC @Homebrew/core folks for thoughts.

**Comment by p-linnane** [comment]
I agee with @MikeMcQuaid thoughr process. @alebcay also suggested this in Slack. 

Given that we have a few supported versions of Go at a time, we should just move to the method we use for Python.

**Comment by SMillerDev** [comment]
I think we've python we've exchanged "why isn't X released" for "why isn't X default" or "why does Y still depend on an old X" and maintainers spend the rest of the year updating the dependencies to the new version.

I am not sure there is a solution that solves this problem besides: - less blocking problems (by regression tests on a regular basis)
- more help (by inviting people complaining about slow updates to help)
- stop packaging statically linked languages that encourage abandonware?

**Comment by alebcay** [comment]
> would just get complaints of "why do I need to have multiple versions of Go installed?"

With `go` being a build dependency, I would hope that this issue is less prevalent than with `python`, where it is a runtime dependency and an end user may have to keep multiple versions around to keep their desired formulae working.

> stop packaging statically linked languages that encourage abandonware?

I think we can be more aggressive about dropping abandoned dependents. I think not packaging the entire language (or software written in a particular language) is throwing the baby out with the bathwater.

By this logic, we should not package Rust, Ocaml, or Haskell, or formulae written in these languages either.

**Comment by p-linnane** [comment]
>By this logic, we should not package Rust, Ocaml, or Haskell, or formulae written in these languages either.

Agreed. Ocaml and Haskell are big offenders here. We still do not have Ocaml 5.0 packaged, and any time GHC has an issue @cho-m is (almost) singlehandedly fixing it. Yet we still ship these, and I think we should continue to do so.

Being aggressive about dropping abandoned dependents and moving Go to a versioned formula make the most sense to me here.

**Comment by johanneswuerbach** [comment]
> maintainers spend the rest of the year updating the dependencies to the new version.

How about (automatically) attempting to bump the go version every time somebody tries to upgrade a formula using go?

I think expecting formulas using go to have released a version with support for go version X directly after version X is out, will always be an uphill battle.

On the flip side, once a new formulae version is pushed, it seems much more likely that this version will also build against the stable version of go at the time.

**Comment by SMillerDev** [comment]
> This would also distribute the load of fixing compatibility issues from a yearly "upgrade fest" to many small events, which have less CI runtime and a much smaller scope to reason about.

Much more likely scenario, the updates go version gets ignored because it's inconvenient for updates or because the formula never gets updates and we still have a yearly upgrade fest but it's in the deprecation pull request so it's sure to get no help other than maintainers.

**Comment by johanneswuerbach** [comment]
> the updates go version gets ignored because it's inconvenient for updates

If a formula uses new go features, the bump can't be ignored and will be necessary as part of the update otherwise this might be a risk, true. Maybe all version updates with an old go version could be rejected after 6 months to ensure people eventually upgrade.

> because the formula never gets updates and we still have a yearly upgrade fest but it's in the deprecation pull request so it's sure to get no help other than maintainers.

If a formula wasn't updated within say, 9 months, can't it just be marked as deprecated and removed with the removal of the old go version?

In case people disagree with this, they can simple do the upgrade and remove the deprecation flag, but it shouldn't be up to the maintainers to ensure everything builds.

**Comment by carlmjohnson** [comment]
> I think expecting formulas using go to have released a version with support for go version X directly after version X is out, will always be an uphill battle.

Yes. In theory the packages should have all been tested by the developer with Go betas during the beta period and be ready to go on day one of release, but that's not the real world. 😆 Much more likely, the package dev is waiting for Homebrew to upgrade before they bother testing building their program with the new Go. 😅 I think it will spread out the work in a more maintainable way to have a deprecation bonanza for programs falling off the back of the support window (which will hopefully be fewer formula) versus having an upgrade bonanza twice year (Go has a new release every six months, so spending ~1 month in upgrade bonanza hell is a significant chunk of the life of a release).

**Comment by billinghamj** [comment]
Has any consideration been given to aggressively chucking out formulae which misbehave?

e.g. a simple (though indeed aggressive) approach could be to:

1. update the runtime version to the latest
2. for any formulae which fail, try those against the previous version
3. for any still-failing formulae, remove them entirely

Then it's entirely up to the upstream maintainers to fix it and resubmit. If they find themselves doing that a lot, it might serve as a motivation to make their formulae more robust

In theory that process could also be fully automated, which could allow for merging with a single CI run (though in practice I realise it'd probably involve a second run for checking against the previous version)

**Comment by Porkepix** [comment]
> If a formula wasn't updated within say, 9 months, can't it just be marked as deprecated and removed with the removal of the old go version?

Some software, while not dead, can be updated less frequently than that, though.

> > stop packaging statically linked languages that encourage abandonware?
> 
> I think we can be more aggressive about dropping abandoned dependents, or have stricter requirements about formula notability. I think not packaging the entire language (or software written in a particular language) is throwing the baby out with the bathwater.
> 
> By this logic, we should not package Rust, Ocaml, or Haskell, or software written in these languages either.

A little bit on and also out of the topic, but because it again crossed my mind not long ago, when I see compiled language releasing a new version and updated here, and that's especially applicable for Go where I've seen this again on the current upgrade: releases can for example include security fixes. But if formulas depending on it doesn't explicitly needs a bump because of their own dependents breaking, bottles won't be rebuilt… and therefore won't benefit from those fixes or whatever else the new language version brings.
Linux package managers tends to more frequent revisions I believe, but that would also means much much more heavy load on CI, and maybe more problems.

**Comment by stefanb** [comment]
Don't forget that build tools that are linking statically (such as Go) only affect future builds, meaning that:
- existing binary builds of old formulae will continue to work identical as before, regardless of the build tool upgrade (until a new build is done)
- when a newer build tool release contains any security fixes all the formulae depending on it should be rebuilt and their new build (revision in homebrew lingo?) released.

Old build tools aren't maintained and updated indefinitely - in case of Go only the last 2 major releases (1.major.minor)  are getting the security fixes.

**Comment by jgimenez** [comment]
I second @billinghamj's proposal of pinning the build tool version, to give everyone some flexibility in the upgrade process,  whilst enforcing that the build tool is in a supported version.

In the case of go, that would be two major versions, meaning at formulas must be updated within 6 months of the latest major release, which I think is a good compromise: https://go.dev/security/policy
